moveit_core/test/test_multi_threaded.cpp

```diff
@@ -0,0 +1,129 @@
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Jens Petit */
+
+#include <moveit/robot_model/robot_model.h>
+#include <moveit/robot_state/robot_state.h>
+#include <moveit/utils/robot_model_test_utils.h>
+#include <moveit/planning_scene/planning_scene.h>
+#include <urdf_parser/urdf_parser.h>
+#include <gtest/gtest.h>
+#include <thread>
+
+#include <moveit/collision_detection/collision_common.h>
+#include <moveit/collision_detection/collision_env.h>
+#include <moveit/collision_detection/collision_detector_allocator.h>
+
+#include <urdf_parser/urdf_parser.h>
+
+const int TRIALS = 1000;
+const int THREADS = 2;
+
+void runCollisionDetection(unsigned int id, unsigned int trials, const planning_scene::PlanningScene* scene,
+                           const robot_state::RobotState* state)
+{
+  collision_detection::CollisionRequest req;
+  for (unsigned int i = 0; i < trials; ++i)
+  {
+    collision_detection::CollisionResult res;
+    scene->checkCollision(req, res, *state);
+  }
+}
+
+class CollisionDetectorThreadedTest : public testing::Test
+{
+protected:
+  void SetUp() override
+  {
+    robot_model_ = moveit::core::loadTestingRobotModel("panda");
+    robot_model_ok_ = static_cast<bool>(robot_model_);
+    robot_state_.reset(new robot_state::RobotState(robot_model_));
+    planning_scene_.reset(new planning_scene::PlanningScene(robot_model_));
+  }
+
+  void TearDown() override
+  {
+  }
+
+  bool robot_model_ok_;
+
+  robot_model::RobotModelPtr robot_model_;
+
+  collision_detection::CollisionEnvPtr cenv_;
+
+  collision_detection::AllowedCollisionMatrixPtr acm_;
+
+  robot_state::RobotStatePtr robot_state_;
+
+  planning_scene::PlanningScenePtr planning_scene_;
+};
+
+TEST_F(CollisionDetectorThreadedTest, InitOk)
+{
+  ASSERT_TRUE(robot_model_ok_);
+}
+
+/** \brief Tests the FCL collision detector in multiple threads. */
+TEST_F(CollisionDetectorThreadedTest, FCLThreaded)
+{
+  std::vector<robot_state::RobotStatePtr> states;
+  std::vector<std::thread*> threads;
+
+  for (unsigned int i = 0; i < THREADS; ++i)
+  {
+    robot_state::RobotState* state = new robot_state::RobotState(planning_scene_->getRobotModel());
+    collision_detection::CollisionRequest req;
+    state->setToRandomPositions();
+    state->update();
+    states.push_back(robot_state::RobotStatePtr(state));
+  }
+
+  for (unsigned int i = 0; i < THREADS; ++i)
+    threads.push_back(
+        new std::thread(std::bind(&runCollisionDetection, i, TRIALS, planning_scene_.get(), states[i].get())));
+
+  for (unsigned int i = 0; i < states.size(); ++i)
+  {
+    threads[i]->join();
+    delete threads[i];
+  }
+  ASSERT_TRUE(true);
```

Comment:
This is a good start to make sure nothing crashes, but I think we should also check that the collision detection calls are correct. Could we check each state when adding it to `states` and make sure each result in the thread is correct?

Reply:
Added this in the latest commit